source/reflect.h2

```diff
@@ -344,12 +351,33 @@ function_declaration: @copyable type =
     has_non_void_return_type     : (this) -> bool = n*.has_non_void_return_type();
 
     unnamed_return_type          : (this) -> std::string = n*.unnamed_return_type_to_string();
+    get_parameters: (this)
+        -> std::vector<object_declaration>
+    = {
+        ret: std::vector<object_declaration> = ();
+        for n*.get_function_parameters() do (d) {
+            _ = ret.emplace_back( d*.declaration*&, this );
+        }
+        return ret;
+    }
 
     is_binary_comparison_function: (this) -> bool = n*.is_binary_comparison_function();
 
     default_to_virtual           : (inout this) = _ = n*.make_function_virtual();
 
     make_virtual                 : (inout this) -> bool = n*.make_function_virtual();
+
+    add_initializer: (move this, source: std::string_view)
+      pre(!has_initializer())
+      pre(parent_is_type())
+    = {
+        prefix := n*.pretty_print_visualize(0);
+        prefix.back() = ' '; // Remove final ';'.
+        p := get_parent().as_type();
+        p.add_member(prefix + source);
+        _ = p;
+        mark_for_removal_from_enclosing_type();
+    }
```

Comment:
Two things:
- I wouldn't use duplicate-and-remove here, because here it's easy to graft a new node into the existing parse tree.
- That also avoids the need to visualize-and-reparse existing code that's already been parsed, which so far I've avoided doing. (I hope visualizing round-trips, but I don't thoroughly test for that yet... I have a test case to exercise it,  but it's not exhaustive and I only occasionally manually try to compile the visualization test case's output to make sure it compiles and I don't compare the before/after trees.)

I would just make `add_initializer` follow the model of `add_member`, which is this:

```
    add_member: (inout this, source: std::string_view)
    = {
        decl := parse_statement(source);
        if !decl || !n*.add_type_member(decl) {
            error( std::string("error attempting to add member:\n") + source );
        }
    }
```

So `add_initializer` would similarly be:

```
    add_initializer: (inout this, source: std::string_view)
    = {
        stmt := parse_statement(source); // only parsing `source`, not the whole thing
        if !stmt || !n*.add_function_initializer(stmt) { // need to add this function
            error( std::string("error attempting to add initializer:\n") + source );
        }
    }
```

I'll push a commit to this PR that makes these changes, and provides the new `add_function_initializer` function besides the existing `add_type_member` function in `parse.h`.

---

Alternatively, using preconditions it could be:

```
    add_initializer: (inout this, source: std::string_view)
      pre(!has_initializer(), "cannot add an initializer to a function that already has one")
      pre(parent_is_type()  , "cannot add an initializer to a function that isn't in a type scope")
    = {
        stmt := parse_statement(source); // only parsing `source`, not the whole thing
        if !stmt || !n*.add_function_initializer(stmt) { // need to add this function
            error( std::string("error attempting to add initializer:\n") + source );
        }
    }
```

I'm a fan of preconditions, but in this case the `parse.h` `add_type_member` function (and so similarly the new `add_function_initializer` function) also acts as a precondition test because it tests those conditions and returns `bool`. Note `add_type_member` pretty much has to check the preconditions internally to navigate the tree correctly, and since it returns `false` if it failed we just check that and emit a single diagnostic, which would be hidden by preconditions.

I did try this precondition version, and found that another drawback was that by default it aborted the compiler which was inelegant. So maybe it's worth using preconditions, but with a `<compiler>` contract_group and installing a log-and-continue error handler, or using `t.require` instead which exists for this. I'll think about it more but maybe not do it right now.

Reply:
I think this has convinced me that both `add_member` and the new`add_intializer` should be using `require` more:

```
    add_member: (inout this, source: std::string_view)
    = {
        decl := parse_statement(source);
        require( decl as bool, 
                 "the provided source string is not a valid statement");
        require( decl*.is_declaration(), 
                 "cannot add a member that is not a declaration");
        require( n*.add_type_member(decl),
                 std::string("error attempting to add member:\n") + source );
    }

    add_initializer: (inout this, source: std::string_view)
    = {
        require( !has_initializer(), 
                 "cannot add an initializer to a function that already has one");
        require( parent_is_type(),
                 "cannot add an initializer to a function that isn't in a type scope");

        stmt := parse_statement(source);
        require( stmt as bool,
                 std::string("cannot add an initializer that is not a valid statement: ") + source);
        require (n*.add_function_initializer(stmt),
                 std::string("error attempting to add initializer:\n") + source);
    }
```

This improves the error message. For now I'll update the PR to go that direction. In the future we can see about integrating `require` with preconditions. Interesting exercise, thanks!

Reply:
Hmm, weird how some of the review comments got interleaved with the parallel commit. These comments were intended to go on the previous commit. Ah well!

Reply:
Thank you for the insightful comments.

One thing these versions don't have is the ability to add a `constexpr` initializer
(some relevant discussions include <https://github.com/hsutter/cppfront/issues/387#declaration> and #761).

For using contracts, #751 will be necessary.